corehq/apps/hqwebapp/decorators.py

```diff
@@ -236,6 +237,22 @@ def _inner(request, *args, **kwargs):
     return _inner
 
 
+def setup_two_factor(view_func):
+    """
+    Use this decorator on the dispatch method of a TemplateView subclass
+
+    We use custom forms for two factor setup and authentication
+    This decorate should be use on any view that depends on two factor
+    methods of authentication.
+    """
+    @wraps(view_func)
+    def _inner(view, request, *args, **kwargs):
+        user = request.couch_user if hasattr(request, 'couch_user') else None
+        register_two_factor_methods(user)
```

Comment:
Why do these need to be registered on every request? (Nothing I can see in the registration code appears to reference the view.) Could they be registered once on Django startup, maybe in an app ready handler?

Reply:
Yeah this may have been an oversight on my part. Related to your next comment about removing the allow_phone_2fa logic, at one point I had the impression that I had to fully rely on registering/unregistering methods, but it seems quite possible and ideal that we could register this once and keep our logic to determine which methods are available to the user. I'll look into this.

Reply:
Fixed in 4ccbb9f9c74ee16ae5179cd22299b53e1dbdcabf